### ISSUE TITLE
Squeezing a 4D Sparse Tensor to 3D Sparse Tensor

### ISSUE THREAD

**Issue by imransai** [StanfordVL/MinkowskiEngine#81]
Hello!

Let me explain my problem first using dense tensor in pytorch. I have a B x 32 x D X H x W tensor
which I obtained after some standard 3D convolutions. I want to do a standard 2D convolution by reshaping
the tensor to B x (32 x D) x H x W. It is easy to do it using standard pytorch tools like view and squeeze operations. There are two parts of my question. 
Part 1:
How do I achieve this using sparse convolution tools all thrroughout the network, if at all possible?
Part 2:
The most naive way to do is by converting the sparse tensor to dense tensor, make it contiguous,
reshape the  dimension and apply standard 2D convolution afterwards. The question then is,
a. Will it impede my backpropagation flow?
b. Will it make the training slower if I switch back and forth from sparse convolution to dense convolutions, or am I expecting training time similar to Minowski convolutions maintained 
  all throughout the network?

**Comment by chrischoy** [comment]
Part 1:

If you use Minkowski[Convolution,Pooling] with stride=[32, 1, 1], assuming D is the first axis, you can collapse the features. This might be closest to what you want and if the features are sparse along the depth axis, pooling will be faster than converting the tensor to a dense tensor. For convolution, you have to benchmark to see which is faster, but I assume that the dimensions are very small, so probably there would be not much difference.

Part 2.

a. It does not impede backprop. It would work as you expect.
b. Switching is highly dependent on the sparsity pattern. If the feature map you get is not particularly sparse, then using dense convolution would be faster.

In general, if the spatial dimensions are small, using dense convolution would be faster. Using sparse tensor requires some overhead to keep the coordinates and kernel map; the overhead will pay off only if the space is large enough and data is sparse.

**Comment by imransai** [comment]
Thanks Chris! 
Regarding the question of converting from sparse to dense, which one do you recommend? Using the sparseTensor.dense() property, or decomposing the coordinates and features and make a dense grid from these coordinates and features?

**Comment by chrischoy** [comment]
I just realized I did not update the `dense()` function to use the batch indices on the first column, so I just fixed it on https://github.com/StanfordVL/MinkowskiEngine/commit/aac143c6062001ea9d18c22e0d2536ed078e0125


1. `SparseTensor.dense()` would be faster than making a dense tensor with decomposed coordinates and features.

2. Ah sorry, what I meant on my previous comment was

B x 32 x D x H x W -> Conv with in channel=32, out_channel=64 kernel_size=[D, 1, 1], stride_size=[D, 1, 1] -> B x 64 x 1 x H x W -> dense() -> B x 64 x 1 x H x W -> .squeeze(2) -> B x 64 x H x W

**Comment by imransai** [comment]
Oh okay, thanks Chris!. I just pulled it from the master. Just a naive question, do I need to install it again in order for the changes to take effect, or just pulling it should be fine?

**Comment by chrischoy** [comment]
You would need to install it again.

You do not need to build all objects from scratch. Simply type

`python setup.py install --keep_objs` in the engine root directory.

**Comment by imransai** [comment]
Excellent! Thanks again Chris! I will let you know if I run into further issues.
